Neos.Flow/Classes/Mvc/Controller/ActionController.php

```diff
@@ -624,14 +631,14 @@ public static function getPublicActionMethods($objectManager)
      * @return ViewInterface the resolved view
      * @throws ViewNotFoundException if no view can be resolved
      */
-    protected function resolveView()
+    protected function resolveView(ActionRequest $request)
```

Comment:
i think you explained, but i forgot. Why do we pass in here the $request explicitly, while at other places we use `$this->request` and `$this->response`?

Reply:
I ultimately want to get rid of `$this->request` and `$this->response` so took steps were I felt it is not too breaky. We can go over this and decide were we want to make that boundary and were not. I know it looks a bit hack right now, but I think whatever we already dare to clamp down right now makes sense to do as it eases potential refactoring (like making everything views related a separate service).

Reply:
Thats a good plan.

Im just looking where the variables are still used. And they leak also into other object. So a lot of things are stateful:

- $this->request
- $this->response
- $this->controllerContext
- $this->uriBuilder
- $this->arguments

with that many arguments, its just a question of time until one of our utility traits like `SpecialResponsesSupport` accesses - accidentally e.g. in a bugfix - any of them and were doomed.

Maybe we should take reinforcements like making the trait's methods static?

Reply:
The AbstractController uses this state in

- getControllerContext
- addFlashMessage (needs the request for the storage)
- forward as base action to merge the new params onto
- redirect, a resetted uribuilder and to set a default format based on the current request
- throwStatus, to with the current response to use the body if empty but overrule its status
- mapRequestArgumentsToControllerArguments sets the $this->arguments (can be solved easily because should only be called in `processRequest`


ActionController

- callActionMethod which doesnt make use of the passed parameters??
- handleTargetNotFoundError idk what this is
- addErrorFlashMessage uses the controller context
- forwardToReferringRequest uses the request->getReferringRequest
- getFlattenedValidationErrorMessage


And i also noticed that "replacing" the `$this->request` is probably not a good idea and should be disallowed via readonly? Mutating by object reference is the only way allowed.

Reply:
My experiment to make them final: https://github.com/neos/flow-development-collection/pull/3283/commits/72700b56d4671544dc1bc6858c35717b21374eec

There was only one place where we ourselves replace the variable for real, but i solved it with a `mergeIntoParent`.

Reply:
But as i feared that means that controllers cant be singletons anymore (why are they declared as such even in general???)

The only usecase of a controller being a singleton would be for forwarded requests, but that is hacky as hell to then rely on internal state :D

Reply:
Would be cool to just put the request and reponse into the Controller via constructor and then process().

Reply:
I would actually just like to not provide a response at all.... I don#t know any use case were we put anytihng but an empty response in there.

Reply:
Discussion moved here https://github.com/neos/flow-development-collection/pull/3293